src/main/java/build/buf/protovalidate/Format.java

```diff
@@ -93,153 +92,275 @@ static String format(String fmtString, ListT list) {
           break;
         case 's':
-          formatString(builder, arg);
+          builder.append(formatString(arg));
           break;
         case 'e':
+          builder.append(formatExponential(arg, precision));
+          break;
         case 'f':
+          builder.append(formatFloat(arg, precision));
+          break;
         case 'b':
+          builder.append(formatBinary(arg));
+          break;
         case 'o':
+          builder.append(formatOctal(arg));
+          break;
         default:
-          throw new ErrException("format: unparsable format specifier %s", c);
+          throw new ErrException(
+              "could not parse formatting clause: unrecognized formatting clause \"" + c + "\"");
       }
     }
     return builder.toString();
   }
 
-  /**
-   * Converts a byte array to a hexadecimal string representation.
-   *
-   * @param bytes the byte array to convert.
-   * @param digits the array of hexadecimal digits.
-   * @return the hexadecimal string representation.
-   */
-  private static String bytesToHex(byte[] bytes, char[] digits) {
-    char[] hexChars = new char[bytes.length * 2];
-    for (int j = 0; j < bytes.length; j++) {
-      int v = bytes[j] & 0xFF;
-      hexChars[j * 2] = digits[v >>> 4];
-      hexChars[j * 2 + 1] = digits[v & 0x0F];
-    }
-    return new String(hexChars);
-  }
-
   /**
    * Formats a string value.
    *
-   * @param builder the StringBuilder to append the formatted string to.
    * @param val the value to format.
    */
-  private static void formatString(StringBuilder builder, Val val) {
+  private static String formatString(Val val) {
     TypeEnum type = val.type().typeEnum();
-    if (type == TypeEnum.Bool) {
-      builder.append(val.booleanValue());
-    } else if (type == TypeEnum.String || type == TypeEnum.Int || type == TypeEnum.Uint) {
-      builder.append(val.value());
-    } else if (type == TypeEnum.Bytes) {
-      builder.append(new String((byte[]) val.value(), StandardCharsets.UTF_8));
-    } else if (type == TypeEnum.Double) {
-      formatDecimal(builder, val);
-    } else if (type == TypeEnum.Duration) {
-      formatDuration(builder, val);
-    } else if (type == TypeEnum.Timestamp) {
-      formatTimestamp(builder, val);
-    } else if (type == TypeEnum.List) {
-      formatList(builder, val);
-    } else if (type == TypeEnum.Map) {
-      throw new ErrException("unimplemented string map type");
-    } else if (type == TypeEnum.Null) {
-      throw new ErrException("unimplemented string null type");
+    switch (type) {
+      case Bool:
+        return Boolean.toString(val.booleanValue());
+      case String:
+        return val.value().toString();
+      case Int:
+      case Uint:
+        Optional<String> str = validateNumber(val);
+        if (str.isPresent()) {
+          return str.get();
+        }
+        return val.value().toString();
+      case Bytes:
+        return new String((byte[]) val.value(), StandardCharsets.UTF_8);
```

Comment:
What does this do if the value is invalid UTF-8 here?

Reply:
Talked offline, but in the interest of visibility:

Currently this will print a placeholder �, but the [strings.md](https://github.com/google/cel-spec/blob/master/doc/extensions/strings.md) doc on the cel-spec site specifies:

> The value is formatted as if string(value) was performed and any invalid UTF-8 sequences are replaced with \ufffd. Multiple adjacent invalid UTF-8 sequences must be replaced with a single \ufffd.

So, this is currently non-conformant, but there are no actual conformance tests yet to cover this. Will implement this in a follow-up PR if that's cool.